src/System.Management.Automation/DscSupport/JsonDscClassCache.cs

```diff
@@ -0,0 +1,3249 @@
+            {
+                return;
+            }
+
+            // keyword is already defined and we don't allow redefine it
+            if (!CacheResourcesFromMultipleModuleVersions && DynamicKeyword.ContainsKeyword(keyword.Keyword))
+            {
+                var oldKeyword = DynamicKeyword.GetKeyword(keyword.Keyword);
+                if (oldKeyword.ImplementingModule == null ||
+                    !oldKeyword.ImplementingModule.Equals(moduleName, StringComparison.OrdinalIgnoreCase) || oldKeyword.ImplementingModuleVersion != moduleVersion)
+                {
+                    var e = PSTraceSource.NewInvalidOperationException(ParserStrings.DuplicateKeywordDefinition, keyword.Keyword);
+                    e.SetErrorId("DuplicateKeywordDefinition");
+                    throw e;
+                }
+            }
+            // Add the dynamic keyword to the table
+            DynamicKeyword.AddKeyword(keyword);
+
+            // And now define the driver functions in the current scope...
+            if (functionsToDefine != null)
+            {
+                functionsToDefine[moduleName + "\\" + keyword.Keyword] = CimKeywordImplementationFunction;
+            }
+        }
+
+        /// <summary>
+        /// A method to generate a keyword from a CIM class object. This is used for DSC.
+        /// </summary>
+        /// <param name="moduleName"></param>
+        /// <param name="moduleVersion"></param>
+        /// <param name="cimClass"></param>
+        /// <param name="runAsBehavior">To specify RunAs behavior of the class.</param>
+        private static DynamicKeyword CreateKeywordFromCimClass(string moduleName, Version moduleVersion, dynamic cimClass, DSCResourceRunAsCredential runAsBehavior)
+        {
+            var resourceName = cimClass.ClassName;
+            string alias = GetFriendlyName(cimClass);
+            var keywordString = string.IsNullOrEmpty(alias) ? resourceName : alias;
+
+            //
+            // Skip all of the base, meta, registration and other classes that are not intended to be used directly by a script author
+            //
+            if (System.Text.RegularExpressions.Regex.Match(keywordString, "^OMI_Base|^OMI_.*Registration", System.Text.RegularExpressions.RegexOptions.IgnoreCase).Success)
+            {
+                return null;
+            }
+
+            var keyword = new DynamicKeyword()
+            {
+                BodyMode = DynamicKeywordBodyMode.Hashtable,
+                Keyword = keywordString,
+                ResourceName = resourceName,
+                ImplementingModule = moduleName,
+                ImplementingModuleVersion = moduleVersion,
+                SemanticCheck = CheckMandatoryPropertiesPresent
+            };
+
+            // If it's one of reserved dynamic keyword, mark it
+            if (System.Text.RegularExpressions.Regex.Match(keywordString, reservedDynamicKeywords, System.Text.RegularExpressions.RegexOptions.IgnoreCase).Success)
```

Comment:
Given that this uses a constant value for the regex, it should be a static compiled regex rather than one created on each call

Reply:
good point; updated.